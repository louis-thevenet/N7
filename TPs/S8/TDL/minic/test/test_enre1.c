/*
Ce programme vérifie la bonne gestion des Enregistrements en MiniC avec des
exemples valides
*/

monProgramme {

    // Déclaration d'un enregistrement (struct)
    typedef struct Personne {
        int age;
        boolean majeur;
    } Personne;

    // Déclaration et initialisation d'une variable de type Personne
    Personne p1 = {21, true};

    // Affectation à partir d'un autre enregistrement
    Personne p2 = p1;

    print p2.age; // 21

    p2.age = p2.age + 1;
    print p2.age; // 22

    // Utilisation de champs dans des expressions
    boolean estMajeur = p1.age >= 18 && p1.majeur;

    print estMajeur; // true

    // Déclaration d'un enregistrement imbriqué
    typedef struct Adresse {
        int numero;
        int codePostal;
    } Adresse;

    typedef struct Client {
        Personne infos;
        Adresse adresse;
    } Client;

    // Utilisation d’un enregistrement imbriqué
    Adresse a1 = {55, 31};
    Client c1 = {p1, a1};

    print a1.numero;    // 55
    print p1.age;       // 21
    print c1.infos.age; // 21
    c1.infos.age = 66;

    int client_age(Client c) { return c.infos.age; }
    int addre_postal(Client c) { return c.adresse.codePostal; }
    print client_age(c1);   // 66
    print addre_postal(c1); // 31
}
