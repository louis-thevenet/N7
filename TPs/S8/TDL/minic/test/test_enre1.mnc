/*
Ce programme vérifie la bonne gestion des Enregistrements en MiniC avec des exemples valides
*/

monProgramme {

    // Déclaration d'un enregistrement (struct)
    typedef struct Personne {
        int age;
        boolean majeur;
    } Personne;

    // Déclaration et initialisation d'une variable de type Personne
    Personne p1;
    p1.age = 20;
    p1.majeur = true;

    // Affectation à partir d'un autre enregistrement
    Personne p2;
    p2 = p1;

    // Modification d’un champ
    p2.age = 17;
    p2.majeur = false;

    // Utilisation de champs dans des expressions
    boolean estMajeur = p1.age >= 18 && p1.majeur;

    // Déclaration d'un enregistrement imbriqué
    typedef struct Adresse {
        int numero;
        int codePostal;
    } Adresse;

    typedef struct Client {
        Personne infos;
        Adresse adresse;
    } Client;

    // Utilisation d’un enregistrement imbriqué
    Client c1;
    c1.infos.age = 30;
    c1.infos.majeur = true;
    c1.adresse.numero = 12;
    c1.adresse.codePostal = 75000;

    // Accès à des champs imbriqués
    int code = c1.adresse.codePostal;
    boolean adulte = c1.infos.majeur;
}
