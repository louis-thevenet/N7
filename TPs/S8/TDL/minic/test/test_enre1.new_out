2122false5521216666
