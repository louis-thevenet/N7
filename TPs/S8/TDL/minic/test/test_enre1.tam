;Assign (seq type) p1
LOADL 21
LOADL 1
;Load variable p1 from offset 0
;Assign (named type) p2
LOADL 0
LOADI (2)
;Load variable p2 from offset 2
;Load field age from record p2, offset 0
;Print:
;p2.age
LOADL 2
LOADL 0
SUBR IAdd
LOADI (1)
SUBR IOut
;Load variable p2 from offset 2
;Load field age from record p2, offset 0
;(p2.age + 1)
;Assignement:
; p2 .age
;from:
;(p2.age + 1)
LOADL 2
LOADL 0
SUBR IAdd
LOADI (1)
LOADL 1
SUBR IAdd
;Assign p2 from SB + 2
;Assign field age from record  p2 , offset 0
LOADL 2
LOADL 0
SUBR IAdd
STOREI (1)
;Load variable p2 from offset 2
;Load field age from record p2, offset 0
;Print:
;p2.age
LOADL 2
LOADL 0
SUBR IAdd
LOADI (1)
SUBR IOut
;Load variable p1 from offset 0
;Load field age from record p1, offset 0
;(p1.age >= 18)
;((p1.age >= 18) && p1.majeur)
LOADL 0
LOADL 0
SUBR IAdd
LOADI (1)
LOADL 18
SUBR IGeq
;Load variable p1 from offset 0
;Load field majeur from record p1, offset 1
LOADL 0
LOADL 1
SUBR IAdd
LOADI (1)
SUBR BAnd
;Load variable estMajeur from offset 4
;Print:
;estMajeur
LOAD (1) 4[SB]
SUBR BOut
;Assign (seq type) a1
LOADL 55
LOADL 31
;Load variable p1 from offset 0
;Assign (seq type) c1
LOADL 0
LOADI (2)
;Load variable a1 from offset 5
LOADL 5
LOADI (2)
;Load variable a1 from offset 5
;Load field numero from record a1, offset 0
;Print:
;a1.numero
LOADL 5
LOADL 0
SUBR IAdd
LOADI (1)
SUBR IOut
;Load variable p1 from offset 0
;Load field age from record p1, offset 0
;Print:
;p1.age
LOADL 0
LOADL 0
SUBR IAdd
LOADI (1)
SUBR IOut
;Load variable c1 from offset 7
;Load field infos from record c1, offset 0
;Load field age from record c1.infos, offset 0
;Print:
;c1.infos.age
LOADL 7
LOADL 0
SUBR IAdd
LOADL 0
SUBR IAdd
LOADI (1)
SUBR IOut
;Assignement:
; c1 .infos.age
;from:
;66
LOADL 66
;Assign c1 from SB + 7
;Assign field infos from record  c1 , offset 0
;Assign field age from record  c1 .infos, offset 0
LOADL 7
LOADL 0
SUBR IAdd
LOADL 0
SUBR IAdd
STOREI (1)
;Function client_age start
JUMP end_client_age
;Load field infos from record c, offset 0
;Load field age from record c.infos, offset 0
client_age
LOADA -4[LB]
LOADL 0
SUBR IAdd
LOADL 0
SUBR IAdd
LOADI (1)
RETURN (1) 0
end_client_age
;Function addre_postal start
JUMP end_addre_postal
;Load field adresse from record c, offset 2
;Load field codePostal from record c.adresse, offset 1
addre_postal
LOADA -4[LB]
LOADL 2
SUBR IAdd
LOADL 1
SUBR IAdd
LOADI (1)
RETURN (1) 0
end_addre_postal
;Print:
;int client_age( Client c ){
;// Return in function : client_age
;return c.infos.age;
;}
;( c1)
CALL (LB) client_age
SUBR IOut
;Print:
;int addre_postal( Client c ){
;// Return in function : addre_postal
;return c.adresse.codePostal;
;}
;( c1)
CALL (LB) addre_postal
SUBR IOut
HALT

