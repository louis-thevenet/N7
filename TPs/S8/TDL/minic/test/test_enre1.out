2122true5521216631
